src/cas.rs

```diff
@@ -1,108 +1,93 @@
+    pub fn set_annotations(&mut self, name: String, annotations: Vec<Annotation>) {
+        self.annotations.insert(name, annotations);
     }
-    pub fn insert_annotation(&mut self, name: &str, annotation: Annotation) {
-        self.annotations
-            .entry(name.to_string())
-            .or_insert_with(|| Vec::new())
-            .push(annotation);
+
+    pub fn insert_annotation(&mut self, name: String, annotation: Annotation) {
+        self.annotations.entry(name).or_default().push(annotation);
     }
-    pub fn print_annotations(&self, name: &str) -> Result<(), PipelineError> {
-        if let Some(annotations) = self.annotations.get(name) {
-            for annot in annotations {
-                let covered_text = self
-                    .get_covered_text_safe(annot.begin as usize, annot.end as usize)
-                    .unwrap();
-                println!(
-                    "ANNOT {} BEGIN {} END {}\nCOVERED TEXT IS: {}",
-                    name, annot.begin, annot.end, covered_text
-                );
-            }
-        } else {
-            return Err(PipelineError::AnnotationMissing);
+
+    pub fn print_annotations(&self, name: &str) -> Result<()> {
+        let annotations = self
+            .annotations
+            .get(name)
+            .ok_or(PipelineError::AnnotationMissing)?;
+
+        for annot in annotations {
+            let covered_text = &self.text[annot.begin..annot.end];
+            println!(
+                "ANNOT {} BEGIN {} END {}\nCOVERED TEXT IS: {}",
+                name, annot.begin, annot.end, covered_text
+            );
         }
+
         Ok(())
     }
+
     pub fn get_covered_annotations(
         &self,
         annot_name: &str,
-    ) -> Result<HashMap<String, Vec<&Annotation>>, PipelineError> {
+    ) -> Result<HashMap<String, Vec<&Annotation>>> {
         let mut covered_annotations: HashMap<String, Vec<&Annotation>> = HashMap::new();
-        if let Some(annot_cover) = self.annotations.get(annot_name) {
-            for annot in annot_cover {
-                let begin = annot.begin;
-                let end = annot.end;
-                for (name, annotations) in &self.annotations {
-                    if name != annot_name {
-                        for a in annotations {
-                            if a.begin >= begin && a.end <= end {
-                                covered_annotations
-                                    .entry(name.to_string())
-                                    .or_insert_with(|| Vec::new())
-                                    .push(a);
-                            }
-                        }
-                    }
+        let annot_cover = self
+            .annotations
+            .get(annot_name)
+            .ok_or(PipelineError::AnnotationMissing)?;
+
+        for annot in annot_cover {
+            for (name, annotations) in &self.annotations {
+                if name != annot_name {
+                    covered_annotations
```

Comment:
This way requires way less lookups and cloning name, since you insert a whole iterator at a time.